django/db/migrations/graph.py

```diff
@@ -108,27 +131,101 @@ def __init__(self):
+        self.nodes[key] = migration
         self.clear_cache()
 
-    def add_dependency(self, migration, child, parent):
+    def add_dummy_node(self, key, origin, error_message):
+        node = DummyNode(key, origin, error_message)
+        self.node_map[key] = node
+        self.nodes[key] = None
+
+    def add_dependency(self, migration, child, parent, skip_validation=False):
+        """
+        This may create dummy nodes if they don't yet exist.
+        If `skip_validation` is set, validate_consistency should be called afterwards.
+        """
         if child not in self.nodes:
-            raise NodeNotFoundError(
-                "Migration %s dependencies reference nonexistent child node %r" % (migration, child),
-                child
+            error_message = (
+                "Migration %s dependencies reference nonexistent"
+                " child node %r" % (migration, child)
             )
+            self.add_dummy_node(child, migration, error_message)
         if parent not in self.nodes:
-            raise NodeNotFoundError(
-                "Migration %s dependencies reference nonexistent parent node %r" % (migration, parent),
-                parent
+            error_message = (
+                "Migration %s dependencies reference nonexistent"
+                " parent node %r" % (migration, parent)
             )
+            self.add_dummy_node(parent, migration, error_message)
         self.node_map[child].add_parent(self.node_map[parent])
         self.node_map[parent].add_child(self.node_map[child])
+        if not skip_validation:
+            self.validate_consistency()
         self.clear_cache()
 
+    def replace_node(self, key, replacement):
+        """
+        If `key` exists, it is removed and any dependencies that were
+        referencing it are changed to reference the `replacement` node
+        instead.
+        """
+        self.nodes.pop(key, None)
+        node = self.node_map.pop(key, None)
+        if node:
+            replacement_node = self.node_map[replacement]
+            for child in node.children:
+                child.parents.remove(node)
+                if child.key != replacement:
+                    replacement_node.add_child(child)
+                    child.add_parent(replacement_node)
+            for parent in node.parents:
+                parent.children.remove(node)
+                if parent.key != replacement:
+                    replacement_node.add_parent(parent)
+                    parent.add_child(replacement_node)
+
+    def remove_replacement_node(self, key, replaced_nodes):
+        """
+        Think of this as the inverse operation to `replace_node`. Kind of.
+        Removes the replacement node `key` and remaps its child nodes to
+        `replaced_nodes` - the list of nodes it would have replaced.
+        """
+        self.nodes.pop(key, None)
+        node = self.node_map.pop(key, None)
+        for child in node.children:
```

Comment:
Sure, can do that. I don't think this can happen from within `build_graph()`, but if some other bit of code were to call this function a second time, then there would be an issue. I'll add a unit test for that.